### PR TITLE
macOS support for in-context signup

### PR DESCRIPTION
**Pull request by alistairjcbrown**
**Reviewer:** @GioSensation 
**Asana:** https://app.asana.com/0/0/1204013406681367/f

## Description
Add in-context signup support to macOS browser

## Steps to test
See task

### CODE REVIEW

File: src/DeviceInterface/AppleDeviceInterface.js [hunk 1]
```diff
@@ -317,16 +368,20 @@ class AppleDeviceInterface extends InterfacePrototype {
...
+                case 'state': {
+                    // Inform that state has changed, but continue polling
+                    // e.g. in-context signup has been dismissed
+                    this.pollingTimeout = setTimeout(() => poll(), 100)
+                    return resolve({stateChange: true})
+                }
                 case 'stop':
                     // Parent wants us to stop polling
-                    resolve(null)
+                    resolve({stop: true})
```
Comment (GioSensation) [inline src/DeviceInterface/AppleDeviceInterface.js:384]: Why was this resolving with `null` before? Do you have any idea?

File: src/Settings.js [hunk 2]
```diff
@@ -158,19 +158,50 @@ export class Settings {
...
+        if (subtype === 'fullName') {
+            return Boolean(this.availableInputTypes.identities?.firstName || this.availableInputTypes.identities?.lastName)
+        }
+
+        if (subtype === 'expiration') {
+            return Boolean(this.availableInputTypes.creditCards?.expirationMonth || this.availableInputTypes.creditCards?.expirationYear)
+        }
+
+        return Boolean(this.availableInputTypes[mainType]?.[subtype])
+    }
```
[the remark this answers is not in the document]
Comment (GioSensation) [reply src/Settings.js:183]: I'd like to sync on this to understand it better. I'll ping you later.

File: src/InContextSignup.js [hunk 3]
```diff
@@ -32,23 +34,50 @@ export class InContextSignup {
...
+            // if the device hasn't provided a clear indication
+            return true
+        }
+    }
+
+    /**
+     * @param {import('./Form/matching.js').SupportedTypes} [inputType]
+     * @returns {boolean}
+     */
+    isAvailable (inputType = 'identities.emailAddress') {
```
Comment (GioSensation) [inline src/InContextSignup.js:56]: From a cursory look it seems that if we use `emailAddress` (without the `identities.` prefix) we could simplify some of the data passing in the callers (no need to import new dependencies on those files and sometimes the piece of data may already be available). Any reason not to?

**Review by GioSensation — COMMENTED** [review]
Overall seems great. I've given this a first look and I'd like to sync later today if you're available. Talk later.

File: src/autofill-utils.js [hunk 4]
```diff
@@ -390,7 +390,7 @@ const tldrs = /\.(?:c(?:o(?:m|op)?|at?|[iykgdmnxruhcfzvl])|o(?:rg|m)|n(?:et?|a(?
  * @returns {boolean}
  */
 function isValidTLD (hostname = window.location.hostname) {
-    return tldrs.test(hostname)
+    return tldrs.test(hostname) || hostname === 'fill.dev'
```
[the remark this answers is not in the document]
Comment (alistairjcbrown) [reply src/autofill-utils.js:393]: @GioSensation That's correct -- it's part of the `isOnExpectedPage` check we do for in-context signup which is there to make sure we're only showing it to the user when it makes the most sense to.

File: src/Settings.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (alistairjcbrown) [reply src/Settings.js:183]: @GioSensation Sounds good. For context, the aim of this function is to tell us "are there saved items showing in the autofill tooltip", which is then used to detect if the in-context signup item is alone in the tooltip or amongst other items.
Hopefully we can find a way to extract the code logic and share it between here and `canAutofillType` 🤞

File: src/InContextSignup.js (hunk 3 above)
Comment (alistairjcbrown) [reply src/InContextSignup.js]: Good question, I think I went down this route because `src/UI/controllers/HTMLTooltipUIController.js` only has the full input type (but of course we can derive the subtype). Not sure it simplifies much (there's still places we need to import), but it keeps us aligned with the current condition `subtype === 'emailAddress' && ...` in `src/Form/inputTypeConfig.js` -- added a commit to change this 👍

File: src/InContextSignup.js [hunk 5]
```diff
@@ -32,23 +34,50 @@ export class InContextSignup {
...
     }
 
-    isAvailable () {
-        const isEnabled = this.device.settings?.featureToggles.emailProtection_incontext_signup
-        const isLoggedIn = this.device.isDeviceSignedIn()
-        return isEnabled && !isLoggedIn && !this.isPermanentlyDismissed() && this.isOnValidDomain() && this.isInstalledRecently
+    isOnAssetsPage () {
+        // Check if on locally hosted macOS autofill page
+        return window.location.href.endsWith('assets/TopAutofill.html')
+    }
```
Comment (GioSensation) [inline src/InContextSignup.js:40]: Can you use `device.globalConfig.isTopFrame` instead of this function?

File: src/Settings.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (GioSensation) [reply src/Settings.js:183]: Yeah, seems like there's a lot of overlap between the two, that's what was puzzling me.

File: src/Form/Form.js [hunk 6]
```diff
@@ -705,7 +726,8 @@ class Form {
 
         const mainType = getInputMainType(input)
         const subtype = getInputSubtype(input)
-        if (await this.device.settings.canAutofillType(mainType, subtype)) {
+        await this.device.settings.populateDataIfNeeded({ mainType, subtype })
+        if (this.device.settings.canAutofillType({ mainType, subtype }, this.device.inContextSignup)) {
```
Comment (alistairjcbrown) [inline src/Form/Form.js:730]: @GioSensation I've pulled the side affect out of `canAutofillType` so that it can be called synchronously, and existing functions have been updated to call the side effect (`populateDataIfNeeded`) beforehand. This should keep the existing functionality the same, while unlocking being able to call `canAutofillType` from the handler. Thoughts?

Comment (GioSensation) [reply src/Form/Form.js:730]: That is AWESOME 🎉! I'm glad I've raised this and you've found a better solution. It was always weird to have an async function to check if we can autofill. One minor note is that I find the name `isUnfillable` confusing. Is it just non-native me? I'll reach out on MM.

File: src/DeviceInterface/AppleDeviceInterface.js [hunk 7]
```diff
@@ -126,16 +131,31 @@ class AppleDeviceInterface extends InterfacePrototype {
      */
     async _show (details) {
         await this._showAutofillParent(details)
-        this._listenForSelectedCredential()
-            .then((response) => {
-                if (!response) {
-                    return
-                }
+        this._listenForSelectedCredential(async (response) => {
```
Comment (alistairjcbrown) [inline src/DeviceInterface/AppleDeviceInterface.js:134]: 🪄  We're using a callback here now instead of a promise so that we can call it multiple times depending on the response, rather than just one resolve with a promise

File: src/InContextSignup.js [hunk 8]
```diff
@@ -22,6 +43,26 @@ export class InContextSignup {
...
+    async openAutofillTooltip () {
+        // Make sure we're working with the latest data
+        await this.device.refreshData()
+
+        // Make sure the tooltip is closed before we try to open it
+        await this.device.uiController?.removeTooltip('stateChange')
+
+        // Make sure the input doesn't have focus so we can focus on it again
+        const activeInput = this.device.activeForm?.activeInput
+        activeInput?.blur()
```
Comment (alistairjcbrown) [inline src/InContextSignup.js:55]: 🪄 This took me longer than I'd like to admit to figure out -- changing tab doesn't cause the input to blur, which means we can't trigger a focus when we come back after signing-up... unless we first make sure to blur it 😅

File: src/DeviceInterface/AppleDeviceInterface.js [hunk 9]
```diff
@@ -304,34 +332,35 @@ class AppleDeviceInterface extends InterfacePrototype {
...
+                return
+            case 'ok': {
+                await callback({data: response.data, configType: response.configType})
+                return
             }
-            poll()
-        })
+            case 'state': {
+                // Inform that state has changed, but continue polling
+                // e.g. in-context signup has been dismissed
```
Comment (GioSensation) [inline src/DeviceInterface/AppleDeviceInterface.js:353]: Why should we keep polling if in-context has been dismissed?

File: src/DeviceInterface/InterfacePrototype.js [hunk 10]
```diff
@@ -576,6 +581,10 @@ class InterfacePrototype {
                 if (this.globalConfig.isDDGDomain && e.data.removeUserData) {
                     this.removeUserData()
                 }
+
+                if (this.globalConfig.isDDGDomain && e.data.closeEmailProtection) {
```
Comment (GioSensation) [inline src/DeviceInterface/InterfacePrototype.js:585]: Not a blocker, but I wonder whether we should rather set the listener only on our domain, as in moving the `this.globalConfig.isDDGDomain` check before we even set the listener.

File: src/Form/Form.js [hunk 11]
```diff
@@ -580,10 +581,30 @@ class Form {
 
     shouldOpenTooltip (e, input) {
         if (!isPotentiallyViewable(input)) return false
-        if (this.device.globalConfig.isApp) return true
-        if (this.device.globalConfig.isWindows) return true
+
+        // Always open if the user has clicked on the Dax icon
         if (isEventWithinDax(e, input)) return true
-        if (this.device.inContextSignup?.isAvailable()) return false
+        if (this.device.globalConfig.isWindows) return true
```
Comment (GioSensation) [inline src/Form/Form.js:587]: Not a blocker, but ideally we'd have an `isDesktopApp` in the globalConfig and here we'd check for: `isDesktopApp && !isIncontextSignupAvailable` or something like that, rather than relying on this being a specific platform.

File: src/Form/matching.js [hunk 12]
```diff
@@ -237,10 +237,12 @@ class Matching {
 
             if (this.subtypeFromMatchers('email', input) && this.isInputLargeEnough('email', input)) {
                 if (opts.isLogin || opts.isHybrid) {
+                    // TODO: Being this support back in the future
+                    // <ASANA LINK HERE>
```
Comment (GioSensation) [inline src/Form/matching.js:241]: Don't forget the link 😉.

File: src/InContextSignup.js [hunk 13]
```diff
@@ -14,6 +16,25 @@ export class InContextSignup {
...
+            Object.defineProperty(window, 'openAutofillAfterClosingEmailProtectionTab', {
+                enumerable: false,
+                configurable: false,
+                writable: false,
+                value: () => {
+                    this.openAutofillTooltip()
+                }
+            })
+        } catch (e) {
+            // Ignore if function can't be set up, it's a UX enhancement not a critical flow
```
Comment (GioSensation) [inline src/InContextSignup.js:36]: Love this little comment.

File: src/InContextSignup.js [hunk 14]
```diff
@@ -32,23 +73,45 @@ export class InContextSignup {
...
     }
 
-    onIncontextSignupDismissed () {
-        this.device.removeAutofillUIFromPage()
+    onIncontextSignupDismissed (options = { shouldHideTooltip: true }) {
+        if (options.shouldHideTooltip) this.device.removeAutofillUIFromPage()
         this.permanentlyDismissedAt = new Date().getTime()
         this.device.deviceApi.notify(new SetIncontextSignupPermanentlyDismissedAtCall({ value: this.permanentlyDismissedAt }))
         this.device.firePixel({pixelName: 'incontext_dismiss_persisted'})
+        if (options.shouldHideTooltip) this.device.deviceApi.notify(new CloseAutofillParentCall(null))
```
Comment (GioSensation) [inline src/InContextSignup.js:111]: Any reason to have this conditional twice?

File: src/Settings.js [hunk 15]
```diff
@@ -158,26 +158,65 @@ export class Settings {
...
+     * }} types
+     * @param {import("./InContextSignup.js").InContextSignup?} inContextSignup
+     * @returns {boolean}
+     */
+    canAutofillType ({mainType, subtype}, inContextSignup) {
+        if (this.isTypeUnavailable({ mainType, subtype })) return false
+
+        // If it's an email field and Email Protection is enabled, return true regardless of other options
+        const isEmailProtectionEnabled = subtype === 'emailAddress' && this.featureToggles.emailProtection && this.availableInputTypes.email
+        if (isEmailProtectionEnabled) return true
```
Comment (GioSensation) [inline src/Settings.js:217]: Nit. I would split it like this, so it's clearer that the two feature flags mean "email protection enabled".

```suggestion
        const isEmailProtectionEnabled = this.featureToggles.emailProtection && this.availableInputTypes.email
        if (subtype === 'emailAddress' && isEmailProtectionEnabled) return true
```

File: src/Settings.test.js [hunk 16]
```diff
@@ -29,8 +35,8 @@ describe('Settings', () => {
                         credentials: {username: true, password: true}
                     })
                 },
-                async (settings) => {
-                    expect(await settings.canAutofillType('credentials', 'password')).toBe(true)
+                (settings) => {
+                    expect(settings.canAutofillType({ mainType: 'credentials', subtype: 'password' }, null)).toBe(true)
```
Comment (GioSensation) [inline src/Settings.test.js:39]: Nice dropping all these async/awaits.

File: src/UI/controllers/HTMLTooltipUIController.js [hunk 17]
```diff
@@ -92,7 +93,9 @@ export class HTMLTooltipUIController extends UIController {
          */
         const tooltipOptions = {
             ...this._htmlTooltipOptions,
-            remove: () => this.removeTooltip()
+            remove: () => this.removeTooltip(),
+            isIncontextSignupAvailable: () => !!this._options.device.inContextSignup?.isAvailable(getSubtypeFromType(topContextData.inputType))
```
Comment (GioSensation) [inline src/UI/controllers/HTMLTooltipUIController.js:97]: That's a handful. Can we get the subtype in a `const` above?

File: src/UI/controllers/OverlayUIController.js [hunk 18]
```diff
@@ -156,9 +156,9 @@ export class OverlayUIController extends UIController {
         }
 
         try {
-            await this._options.show(details)
             this.#state = 'parentShown'
             this._attachListeners()
+            this._options.show(details)
```
[the remark this answers is not in the document]
Comment (GioSensation) [reply src/UI/controllers/OverlayUIController.js:161]: Sounds good, go ahead.

File: integration-test/helpers/harness.js [hunk 19]
```diff
@@ -12,10 +12,17 @@ const DATA_DIR_PREFIX = 'ddg-temp-'
...
+        html: 'text/html',
+        css: 'text/css',
+        default: 'text/plain'
+    }
     await page.route(`${domain}/**/*`, (route, request) => {
         const { pathname } = new URL(request.url())
+        const fileType = pathname.split('.').pop()
         return route.fulfill({
             status: 200,
+            contentType: contentType[fileType] || contentType.default,
```
Comment (GioSensation) [inline integration-test/helpers/harness.js:25]: Mhm, why is this needed? Sorry, I forgot.

**Review by GioSensation — APPROVED** [review]
This is awesome work. It was a pretty complex piece of implementation and it's done brilliantly. I've left just a few nits here and there, but overall it seems grand.

File: src/DeviceInterface/AppleDeviceInterface.js (hunk 9 above)
Comment (alistairjcbrown) [reply src/DeviceInterface/AppleDeviceInterface.js:353]: @GioSensation When there are identities available, dismissing in-context signup doesn't close the tooltip, it just removes the in-context signup option. We'll want to keep polling so that selecting the identity after dismissing works as expected.

File: src/Form/Form.js (hunk 11 above)
Comment (alistairjcbrown) [reply src/Form/Form.js:587]: @GioSensation This is a weird diff, but that line wasn't added in this PR, it's from a few lines up (but +1 to removing platform specific checks)
This PR moved the `isApp` check down and essentially facilitates the check of not opening the tooltip if, in the macOS app, we don't have anything to autofill except in-context signup.

File: src/InContextSignup.js (hunk 14 above)
Comment (alistairjcbrown) [reply src/InContextSignup.js:111]: @GioSensation Just that these two control hiding the tooltip. I wasn't sure if the order of calls mattered (remove from page, trigger persisting dismissal data, close parent). If not, I can combine in a single conditional (probably move top line to the bottom?)

Comment (GioSensation) [reply src/InContextSignup.js:111]: Yes, I'd move to the bottom. It should make no difference.

File: integration-test/helpers/harness.js (hunk 19 above)
Comment (alistairjcbrown) [reply integration-test/helpers/harness.js:25]: @GioSensation I think this is the first time we've used `setupMockedDomain` on Playwright's Safari engine, but on mocked tests the CSS file wasn't loading because the content type wasn't set. This now sets up the correct content type for the mocked response depending on the file extension pulled from the URL.

File: src/UI/styles/autofill-tooltip-styles.css [hunk 20]
```diff
@@ -110,6 +111,37 @@
...
+    margin: 0 0 0 4px;
+    padding: 1px 6px;
+}
+.tooltip__incontext__cta button:first-of-type {
+    margin: 0;
+}
+.tooltip__incontext__cta button.primary {
+    background: linear-gradient(180deg, #1F8AFF 0%, #0173FF 100%);
+    border: 0.5px solid rgba(40, 145, 255, 0.05);
+    box-shadow: 0px 0px 1px rgba(40, 145, 255, 0.05), 0px 1px 1px rgba(40, 145, 255, 0.1);
```
[the remark this answers is not in the document]
Comment (alistairjcbrown) [reply src/UI/styles/autofill-tooltip-styles.css:130]: @GioSensation The border gave a subtle difference to the bottom of the button (but also interfered with the overall button height), and the box-shadow I couldn't see. For simplicity, I've nuked them and we'll see if anything get flagged in design review.
